laboratory-service/src/main/java/com/raf/si/laboratoryservice/service/impl/ReferralServiceImpl.java

```diff
@@ -0,0 +1,99 @@
+
+    public ReferralServiceImpl(ReferralRepository referralRepository, LabWorkOrderRepository labWorkOrderRepository, ReferralMapper referralMapper) {
+        this.referralRepository = referralRepository;
+        this.labWorkOrderRepository = labWorkOrderRepository;
+        this.referralMapper = referralMapper;
+    }
+
+    @Override
+    public ReferralResponse createReferral(CreateReferralRequest createReferralRequest) {
+        Referral referral = referralRepository.save(referralMapper.requestToModel(createReferralRequest));
+        return referralMapper.modelToResponse(referral);
+    }
+
+    @Override
+    public ReferralListResponse referralHistory(UUID lbp, Timestamp dateFrom, Timestamp dateTo, Pageable pageable) {
+        Page<Referral> referralPage = referralRepository.findByLbpAndCreationTimeBetweenAndDeletedFalse(lbp, dateFrom, dateTo, pageable);
+        return referralMapper.referralPageToReferralListResponse(referralPage);
+    }
+    @Override
+    public ReferralResponse getReferral(Long id) {
+        Referral referral = referralRepository.findById(id).orElseThrow(() -> {
+            log.error("Ne postoji uput sa id-ijem '{}'", id);
+            throw new NotFoundException("Uput sa datim id-ijem ne postoji");
+        });
+        return referralMapper.modelToResponse(referral);
+    }
+
+    public ReferralResponse deleteReferral(Long id) {
+        Referral referral = referralRepository.findById(id).orElseThrow(() -> {
+            log.error("Ne postoji uput sa id-ijem '{}'", id);
+            throw new NotFoundException("Uput sa datim id-ijem ne postoji");
+        });
+
+        UUID lbzFromToken = TokenPayloadUtil.getTokenPayload().getLbz();
+        boolean lbzMatch = referral.getLbz().equals(lbzFromToken);
+
+        if (lbzMatch && labWorkOrderRepository.findByReferral(referral) == null) {
+            referral.setDeleted(true);
+            referral = referralRepository.save(referral);
+            log.info("Uput sa id-ijem '{}' je uspesno obrisan", id);
+            return referralMapper.modelToResponse(referral);
+        } else {
+            log.error("Lbz uputa i lbz iz tokena se ne poklapaju '{}'", id);
+            throw new NotFoundException("Uput nije obrisan");
+        }
+    }
+
+    @Override
+    public ReferralListResponse unprocessedReferrals(UUID lbp) {
+        UUID pboFromToken = TokenPayloadUtil.getTokenPayload().getPbo();
+        List<Referral> unprocessedReferralsByThreeParams =
+                referralRepository.findByLbpAndPboReferredToAndStatus(lbp, pboFromToken, ReferralStatus.NEREALIZOVAN).orElseThrow(() -> {
+                    log.error("Ne postoji trazeni uput");
+                    throw new NotFoundException("Uput sa zadatim parametrima nije pronadjen");
+                });
+
+        List<Referral> unprocessedReferrals = unprocessedReferralsByThreeParams.stream()
+                .filter(referral -> labWorkOrderRepository.findByReferral(referral) == null)
```

Comment:
Ne moramo za svaki referral da komuniciramo sa bazom. Mozes to da resis ili sa custom query-jem koji ce da koristi fetch eager pa da za svaki referral proveravas da li mu je polje labWorkOrder (ili kako god da se zove) null, ili da napravis query koji ce da u labworkrepo nadje sve radne naloge ciji se referral nalazi u listi koju si izvukao uz pomoc Where In